rust/arrow/src/array/transform/utils.rs

```diff
@@ -61,3 +61,17 @@ pub(super) fn extend_offsets<T: OffsetSizeTrait>(
         buffer.extend_from_slice(last_offset.to_byte_slice());
     });
 }
+
+#[inline]
+pub(super) unsafe fn get_last_offset<T: OffsetSizeTrait>(
+    offset_buffer: &MutableBuffer,
+) -> T {
+    // JUSTIFICATION
+    //  Benefit
+    //      20% performance improvement extend of variable sized arrays (see bench `mutable_array`)
+    //  Soundness
+    //      * offset buffer is always extended in slices of T and aligned accordingly.
+    //      * Buffer[0] is initialized with one element, 0, and thus `mutable_offsets.len() - 1` is always valid.
+    let offsets = offset_buffer.data().align_to::<T>().1;
```

Comment:
I wonder if this would be a good place to use a debug_assert.


Something like the following to ensure the data was actually aligned as intended

```suggestion
    let (prefix, offsets, suffix) = offset_buffer.data().align_to::<T>();
    debug_assert!(prefix.len() == 0 && suffix.len() == 0);
```

Reply:
Or maybe  something more direct:
```
debug_assert!(*offsets.get_unchecked(offsets.len() - 1) == mutable_offsets[mutable_offsets.len() - 1]);
```

Reply:
We use the first option `(prefix, offsets, suffix)` in the parquet crate. I'd support either option that you're suggesting